refactor: optimize member expression creation in AstSnippet

IWANABETHATGUY opened the pull request:
### TL;DR

Optimize member expression creation in AST snippets by using iterative approach instead of recursion.

###

IWANABETHATGUY commented:
* **#6100** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6100?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6091** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6091?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6091).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/rolldown_ecmascript_utils/src/ast_snippet.rs:
```diff
@@ -93,16 +94,29 @@ impl<'ast> AstSnippet<'ast> {
     names: &[CompactStr],
     span: Span,
   ) -> ast::Expression<'ast> {
```
I wonder if we can use `self.member_expr_or_ident_ref(self.void_zero(), names, span)` here.

sapphi-red commented on crates/rolldown_ecmascript_utils/src/ast_snippet.rs:
```diff
@@ -63,28 +63,29 @@ impl<'ast> AstSnippet<'ast> {
+        ast::Expression::from(self.builder.member_expression_static(
+          SPAN,
+          cur,
+          self.id_name(name, SPAN),
```
I think we have to pass the original span here to preserve the mapping. I expect `import_react` to be mapped from `react` and `version` to be mapped from `version`.
<img width="1638" height="344" alt="image" src="https://github.com/user-attachments/assets/b8648ab0-359b-42b7-ab8c-a6fad370c6f9" />
[REPL](https://repl.rolldown.rs/#eNptj7FuwzAMRH9F4OIUCNR0FdDv6KJFsOlUhUwaEpUGMPTvoW04yJBBw/FO98gFRnALZE5p4H+yPdMYr/avrFMC9845Q68G3mfOYgYcQ01iFk/GcJW5ituFMYVr7nEKszNdpBQJu9Vonhq0M0Qa8P6CeuodEKcNkDH0YsbMk+ns56Y00/nsaX26VlGb2XzvUXvDXCKTLRKylJ8ovycPXx4+jjgntImvJ/20zZSH4CRX1KUOwPP+Q79eXQ6IQrXbXuzFw1bVtOKmuRQEi0B7AM/ldsI=), [sourcemap visualizer](https://evanw.github.io/source-map-visualization/#MjA0MQAvLyNyZWdpb24gcm9sbGRvd246cnVudGltZQ0KdmFyIF9fY3JlYXRlID0gT2JqZWN0LmNyZWF0ZTsNCnZhciBfX2RlZlByb3AgPSBPYmplY3QuZGVmaW5lUHJvcGVydHk7DQp2YXIgX19nZXRPd25Qcm9wRGVzYyA9IE9iamVjdC5nZXRPd25Qcm9wZXJ0eURlc2NyaXB0b3I7DQp2YXIgX19nZXRPd25Qcm9wTmFtZXMgPSBPYmplY3QuZ2V0T3duUHJvcGVydHlOYW1lczsNCnZhciBfX2dldFByb3RvT2YgPSBPYmplY3QuZ2V0UHJvdG90eXBlT2Y7DQp2YXIgX19oYXNPd25Qcm9wID0gT2JqZWN0LnByb3RvdHlwZS5oYXNPd25Qcm9wZXJ0eTsNCnZhciBfX2NvbW1vbkpTID0gKGNiLCBtb2QpID0+IGZ1bmN0aW9uKCkgew0KCXJldHVybiBtb2QgfHwgKDAsIGNiW19fZ2V0T3duUHJvcE5hbWVzKGNiKVswXV0pKChtb2QgPSB7IGV4cG9ydHM6IHt9IH0pLmV4cG9ydHMsIG1vZCksIG1vZC5leHBvcnRzOw0KfTsNCnZhciBfX2NvcHlQcm9wcyA9ICh0bywgZnJvbSwgZXhjZXB0LCBkZXNjKSA9PiB7DQoJaWYgKGZyb20gJiYgdHlwZW9mIGZyb20gPT09ICJvYmplY3QiIHx8IHR5cGVvZiBmcm9tID09PSAiZnVuY3Rpb24iKSBmb3IgKHZhciBrZXlzID0gX19nZXRPd25Qcm9wTmFtZXMoZnJvbSksIGkgPSAwLCBuID0ga2V5cy5sZW5ndGgsIGtleTsgaSA8IG47IGkrKykgew0KCQlrZXkgPSBrZXlzW2ldOw0KCQlpZiAoIV9faGFzT3duUHJvcC5jYWxsKHRvLCBrZXkpICYmIGtleSAhPT0gZXhjZXB0KSBfX2RlZlByb3AodG8sIGtleSwgew0KCQkJZ2V0OiAoKGspID0+IGZyb21ba10pLmJpbmQobnVsbCwga2V5KSwNCgkJCWVudW1lcmFibGU6ICEoZGVzYyA9IF9fZ2V0T3duUHJvcERlc2MoZnJvbSwga2V5KSkgfHwgZGVzYy5lbnVtZXJhYmxlDQoJCX0pOw0KCX0NCglyZXR1cm4gdG87DQp9Ow0KdmFyIF9fdG9FU00gPSAobW9kLCBpc05vZGVNb2RlLCB0YXJnZXQpID0+ICh0YXJnZXQgPSBtb2QgIT0gbnVsbCA/IF9fY3JlYXRlKF9fZ2V0UHJvdG9PZihtb2QpKSA6IHt9LCBfX2NvcHlQcm9wcyhpc05vZGVNb2RlIHx8ICFtb2QgfHwgIW1vZC5fX2VzTW9kdWxlID8gX19kZWZQcm9wKHRhcmdldCwgImRlZmF1bHQiLCB7DQoJdmFsdWU6IG1vZCwNCgllbnVtZXJhYmxlOiB0cnVlDQp9KSA6IHRhcmdldCwgbW9kKSk7DQoNCi8vI2VuZHJlZ2lvbg0KLy8jcmVnaW9uIHJlYWN0LmpzDQp2YXIgcmVxdWlyZV9yZWFjdCA9IC8qIEBfX1BVUkVfXyAqLyBfX2NvbW1vbkpTKHsgInJlYWN0LmpzIjogKChleHBvcnRzKSA9PiB7DQoJZXhwb3J0cy52ZXJzaW9uID0gIjEuMC4wIjsNCn0pIH0pOw0KDQovLyNlbmRyZWdpb24NCi8vI3JlZ2lvbiBpbmRleC5qcw0KdmFyIGltcG9ydF9yZWFjdCA9IC8qIEBfX1BVUkVfXyAqLyBfX3RvRVNNKHJlcXVpcmVfcmVhY3QoKSk7DQpjb25zdCBmb28gPSBpbXBvcnRfcmVhY3QudmVyc2lvbi5zdGFydHNXaXRoKCIxIik7DQpjb25zb2xlLmxvZyhmb28pOw0KDQovLyNlbmRyZWdpb24NCi8vIyBzb3VyY2VNYXBwaW5nVVJMPWRhdGE6YXBwbGljYXRpb24vanNvbjtjaGFyc2V0PXV0Zi04O2Jhc2U2NCxleUoyWlhKemFXOXVJam96TENKbWFXeGxJam9pYVc1a1pYZ3Vhbk1pTENKdVlXMWxjeUk2V3lKeVpXRmpkQzUyWlhKemFXOXVMbk4wWVhKMGMxZHBkR2dpWFN3aWMyOTFjbU5sY3lJNld5SXVMaTl5WldGamRDNXFjeUlzSWk0dUwybHVaR1Y0TG1weklsMHNJbk52ZFhKalpYTkRiMjUwWlc1MElqcGJJbVY0Y0c5eWRITXVkbVZ5YzJsdmJpQTlJRndpTVM0d0xqQmNJbHh5WEc0aUxDSnBiWEJ2Y25RZ2NtVmhZM1FnWm5KdmJTQW5MaTl5WldGamRDNXFjeWRjY2x4dVhISmNibU52Ym5OMElHWnZieUE5SUhKbFlXTjBMblpsY25OcGIyNHVjM1JoY25SelYybDBhQ2hjSWpGY0lpbGNjbHh1WTI5dWMyOXNaUzVzYjJjb1ptOXZLVnh5WEc0aVhTd2liV0Z3Y0dsdVozTWlPaUk3T3pzN096czdPenM3T3pzN096czdPenM3T3pzN096czdPenM3UVVGQlFTeFRRVUZSTEZWQlFWVTdPenM3T3p0QlEwVnNRaXhOUVVGTkxHMUNRVUZOUVN4dFFrRkJlVUlzU1VGQlNUdEJRVU42UXl4UlFVRlJMRWxCUVVrc1NVRkJTU0o5MzgxAHsidmVyc2lvbiI6MywiZmlsZSI6ImluZGV4LmpzIiwibmFtZXMiOlsicmVhY3QudmVyc2lvbi5zdGFydHNXaXRoIl0sInNvdXJjZXMiOlsiLi4vcmVhY3QuanMiLCIuLi9pbmRleC5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnRzLnZlcnNpb24gPSBcIjEuMC4wXCJcclxuIiwiaW1wb3J0IHJlYWN0IGZyb20gJy4vcmVhY3QuanMnXHJcblxyXG5jb25zdCBmb28gPSByZWFjdC52ZXJzaW9uLnN0YXJ0c1dpdGgoXCIxXCIpXHJcbmNvbnNvbGUubG9nKGZvbylcclxuIl0sIm1hcHBpbmdzIjoiOzs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7O0FBQUEsU0FBUSxVQUFVOzs7Ozs7QUNFbEIsTUFBTSxtQkFBTUEsbUJBQXlCLElBQUk7QUFDekMsUUFBUSxJQUFJLElBQUkifQ==)
(Not a blocker for this PR)

IWANABETHATGUY replied:
nice suggestion

sapphi-red commented on crates/rolldown_ecmascript_utils/src/ast_snippet.rs:
```diff
@@ -62,47 +62,40 @@ impl<'ast> AstSnippet<'ast> {
-        false,
-      ))
-    }
+    self.member_expr_or_ident_ref(self.void_zero(), names, span, 1)
```
I guess we can pass a trimmed `names` here and remove `cursor`.
```suggestion
    self.member_expr_or_ident_ref(self.void_zero(), &names[1..], span)
```

sapphi-red commented on crates/rolldown_ecmascript_utils/src/ast_snippet.rs:
```diff
@@ -62,47 +62,40 @@ impl<'ast> AstSnippet<'ast> {
-      }
+    let mut cur = object;
+    while cursor < names.len() {
+      let name = names[cursor].as_str();
```
With the change above, I guess we can change this into a for-loop (which may reduce the boundary check).
```suggestion
    for name in names {
      let name = name.as_str();
```

IWANABETHATGUY commented:
### Merge activity

* **Sep 9, 11:45 AM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Sep 9, 11:58 AM UTC**: `IWANABETHATGUY` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Sep 9, 12:21 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

IWANABETHATGUY committed:
refactor: optimize member expression creation in AstSnippet (#6091)

### TL;DR

Optimize member expression creation in AST snippets by using iterative approach instead of recursion.

###
